### ISSUE TITLE
dynamically calculate SWAP_EXPIRY_HOURS

### ISSUE THREAD

**Issue by aspiers** [swapmyvote/swapmyvote#590]
We have a Heroku Scheduler add-on which runs `rake swaps:cancel_old` every 60 mins.  The idea of this is that if someone takes too long to respond to a swap offer, the offer gets automatically expired.

The expiry threshold (not to be confused with the hourly frequency with which the scheduler runs) is currently controlled by the `SWAP_EXPIRY_HOURS` environment variable.  However if swapping is open for weeks before election day, then the threshold should be higher (e.g. 24 or 48 hours, and then gradually get reduced as we get closer to election day.  On election day itself, the threshold should be set to maybe ~3 hours in the morning, and then ~1 hour in the afternoon, or something like that.

So it would be good to replace `SWAP_EXPIRY_HOURS` with a more dynamic mechanism.  However it should be noticed that currently we assume this is a static value, and it is substituted within the FAQ, as well as other places.  So there may need to be copy updates to reflect its new dynamic nature.

**Comment by aspiers** [comment]
@tdg Please let me know what you think about this?

**Comment by baob** [comment]
> currently we assume this is a static value, and it is substituted within the FAQ, as well as other places

I was going to dive in and encapsulate this as a function in preparation for this, but it looks to me like it's already been done.

All that's needed is for you guys to agree an algorithm. It can then be implemented in a single place so far as I can see.

**Comment by aspiers** [comment]
I don't think we need an algorithm, just a mapping table with thresholds.  Then the table could have entries reflecting something like the following, with the first matching entry being the one which gets used:

<table>
<tr>
	<th>If we're before this time
	<th>Set expiry to this
<tr>
	<td>7 days before polls open
	<td>48
<tr>
	<td>5 days before polls open
	<td>36
<tr>
	<td>4 days before polls open
	<td>24
<tr>
	<td>3 days before polls open
	<td>18
<tr>
	<td>2 days before polls open
	<td>12
<tr>
	<td>1 day before polls open
	<td>8
<tr>
	<td>less than 1 day before polls open
	<td>6
<tr>
	<td>before 10am on election day
	<td>4
<tr>
	<td>before 12pm on election day
	<td>2
<tr>
	<td>before 3pm on election day
	<td>1
<tr>
	<td>before 6pm on election day
	<td>30 mins
</table>

@tdg What do you think?

**Comment by tdg** [comment]
Exactly and that’s pretty much perfect as a draft timetable too. In 2019 there was demand to speed it up but I actually think some of the other things we’ve done will help with conversions into confirmed swaps too, so this is a great starting point.

**Comment by aspiers** [comment]
Great, thanks.  In terms of code, the above table could be represented as follows:

```ruby
SWAP_EXPIRY_THRESHOLDS = [
    [-7 * 24, 48],
    [-5 * 24, 36],
    [-4 * 24, 24],
    [-3 * 24, 18],
    [-2 * 24, 12],
    [-24, 8],
    [0, 6],
    [10, 4],
    [12, 2],
    [15, 1],
    [18, 0.5],
]
```

and hardcoded into the code-base.  Then it should be very easy to implement.  @baob What do you think?

**Comment by baob** [comment]
> I don't think we need an algorithm, just a mapping table with thresholds.

a mapping table with thresholds defines the algorithm, though,  no  ?   ;)

**Comment by baob** [comment]
Observations:

Since this is based on days till the election we will need an environment variable to tell us what that is, e.g. ELECTION_DATE

This will be very inflexible by comparison to the current solution and so I suggest we leave SWAP_EXPIRY_HOURS in the code as an optional override for the algorithm above. 

This will be hard for anyone other than a coder to verify correctness, and it may therefore take a long time to get it live after coding is finished. Given the election is about 8 days away, will we be able to get it live in time for it to bring any benefit ?

Thoughts  @tdg @aspiers ?

**Comment by tdg** [comment]
Agree, this isn’t priority 

>
